### DISCUSSION TITLE
Discussion: Add canonical status for all blocks in archive node

### DISCUSSION THREAD

**Discussion by garethtdavies** [MinaProtocol/mina#12015] [General]
Anyone working with the archive node has to deal with recursive queries to determine the canonical chain. The archive node now processes blocks older than 290, and they are marked as canonical / orphaned but could this be implemented for all blocks up to the current chain tip?

Clearly, any blocks less than 290 from the current height are subject to a reorg, and so there would be overhead in implementing this - but this could potentially dramatically improve the usability of the archive node if blocks had a canonical state without invoking a recursive query. It would also make integrating it into services that expose the archive data easier to deliver.

**Comment by duccas** [comment]
I fully support the proposal. We also consider the implemented function of determining the canonical chain in the archive node to be absolutely inapplicable.
This was our first and most important task when creating the explorer, because we chose the option of using the archive node and not GraphQL.
And as Gareth said, this requires a rather complex recursive query which wastes resources. 290 blocks detection delay is too big and no one will use it.

**Comment by psteckler** [comment]
How much overhead do you see from performing a recursive query? Hard numbers, if you have them.

**Reply by garethtdavies** [reply]
> How much overhead do you see from performing a recursive query? Hard numbers, if you have them.

Ducca can probably give you performance stats but I was coming at it more from the development overhead than the performance overhead. Every developer who wants to work with the archive node needs to reimplement their own queries, which are non-trivial. 

As a more concrete example, I was playing around with exposing data on Hasura to build a simple API for Berkely to get events/sequence events. The only way to do this currently (due to the recursive queries) is via a Postgres function - I had implemented something like the below to create a table of canonical blocks that I could use. Maybe there are other ways of doing it, and we just need better documentation but simply being able to query the status of the chain (as we can do past 290 blocks) seems a simpler approach to anyone developing atop the archive node.

```
CREATE
OR REPLACE FUNCTION public.table_canonical_blocks() RETURNS SETOF canonical_blocks LANGUAGE sql STABLE AS $ function $ WITH RECURSIVE chain AS (
  (
    SELECT
      id,
      state_hash,
      parent_id,
      creator_id,
      height
    FROM
      blocks b
    WHERE
      height = (
        select
          MAX(height)
        from
          blocks
      )
    ORDER BY
      timestamp ASC
    LIMIT
      1
  )
  UNION ALL
  SELECT
    b.id,
    b.state_hash,
    b.parent_id,
    b.creator_id,
    b.height
  FROM
    blocks b
    INNER JOIN chain ON b.id = chain.parent_id
    AND chain.id <> chain.parent_id
)
SELECT
  c.id,
  c.state_hash,
  c.height
FROM
  chain c
  INNER JOIN public_keys pk ON pk.id = c.creator_id $ function $
```

**Reply by psteckler** [reply]
You can, of course, exclude blocks marked as `orphaned` in your queries, and include all other blocks, without need for a recursive query. You may get some blocks that will be marked `orphaned` in the future.

**Reply by garethtdavies** [reply]
Surely, most, if not all, applications are interested in the canonicity of recent blocks? Can you imagine an application that would be built atop the archive node without using a recursive query?

**Reply by psteckler** [reply]
Well, any recent block will have a path back to a canonical block (I believe). You can check that with a recursive query, or simply assume it. You can consider such blocks as canonical, if you'd like.

**Reply by garethtdavies** [reply]
I'm not sure I'm following. If I want to get some information from a recent block, e.g., some sequence events that were emitted, I need to check the canonicity of the block that included it. As in, I likely only want to do something if the block is in the canonical chain and can't assume it is (and don't want to wait for 290 confirmations). I know I can get this via a recursive query, but it's not simple to implement as opposed to querying against the `chain_status` field. Clearly, there was a rationale to include this column for blocks older than 290 blocks, so the idea is to include this status up to the chain tip (knowing it can be re-orged).

**Reply by psteckler** [reply]
I'm suggesting:
- look at the maximum height
- for blocks at least 290 before that, include if `chain_status` is `canonical`
- for more recent blocks, include them regardless of `chain_status` (the status will always be `pending`)

**Reply by garethtdavies** [reply]
Do you have a sample query as I am not following this.

**Reply by psteckler** [reply]
```
SELECT * FROM blocks
WHERE height > (SELECT MAX(height) FROM blocks) - 290
OR chain_status = 'canonical'
```

**Reply by garethtdavies** [reply]
Right, but I'm still then in a position where I don't know the canonical chain for the top 290 blocks, so have to resolve that with a recursive query (hence this discussion).

That query certainly does help performance wise though 👍

**Reply by psteckler** [reply]
Oh, you want the chain itself, rather than the chain status of those recent blocks?

**Reply by garethtdavies** [reply]
I guess we are going around in circles here a little bit. To rephrase; There is a canonical chain prior to the chain being finalized, as represented by `bestChain` in GraphQL. My suggestion is to include this status in the archive node and not as pending but as the status of the chain currently is (and it can be re-orged). 

The rationale is doing so removes a large amount of overhead for developing atop the archive node, which is currently heavily dependent on making these recursive queries. I appreciate this isn't that simple to implement, but if it drastically eases the end-user development, it could be worth the overhead.

**Reply by psteckler** [reply]
The `bestChain` query relies on the transition frontier in a running daemon. I suppose we could use a similar algorithm in the archive db, marking some blocks as `best_tip` instead of `pending`.

**Reply by garethtdavies** [reply]
Yeh, I meant that just as an example, but we have the data as can walk back via the previous state hash - the tricky bit would be getting the current tip, but feasibly you could use the same tie-breaker rules in the archive node to choose the current tip (compare VRF and hash of block state hash if multiple blocks same producer). My suggestion would be to mark them as canonical/orphaned as they currently are, knowing that they could be re-orged if they are less than 290 blocks deep. I'm struggling to think of any real-world application that actually uses this 290 block finality rather than, say a few confirmations.

**Reply by garethtdavies** [reply]
I suppose marking as `best_tip` would also work if really want to keep the distinction of finalized blocks as a simple `chain_status = 'canonical' OR  chain_status = 'best_tip'` would work and remove the need for a recursive query.

**Reply by jenr24G** [reply]
Hi! I've written a rust program that builds a graph of blocks within the transition frontier, and I'm working on applying various heuristics to the graph to determine what is most likely the canonical chain. I don't think it will be possible to say for 100% sure what the canonical chain is, but my program will be able to give a high level of certainty, and an eventual goal is to be able to give a probablility for a certain block either being canonical or not. I think this solution addresses the underlying problem of this discussion, and I'm wondering how people see this as connecting to the existing archive node solution.

**Reply by garethtdavies** [reply]
I might be misunderstanding this without looking at it, but we have an algorithm, so all block producers agree on the same current tip, roughly:

* Select the longest chain
* If tied, i.e., multiple blocks proposed at the same height, select the "best" VRF of the block winner
* If multiple blocks with the same VRF, i.e., multiple blocks from the same producer, compare the hash of the state hash and select the "best".

My suggestion would be to replicate this in the archive node, then once we know the current tip, we can simply walk back the chain using the previous state hash, updating any pending blocks. Nothing is finalized until 290 blocks have been built atop it, but the probability of a block being canonical will simply be based on its number of confirmations. As per this page https://docs.minaprotocol.com/node-operators/lifecycle-of-a-payment:


| Finality (in blocks) | Average time for finality | Finality confidence (%) |
| -------------------- | ------------------------- | ----------------------- |
| 8                    | 33 mins                   | 98.6709                 |
| 15                   | 60 mins                   | 99.9231                 |
| 23                   | 1hr 32mins                | 99.9965                 |
| 30                   | 2hrs                      | 99.9998                 |
| 38                   | 2hrs 32mins               | 100                     |

**Reply by jenr24G** [reply]
My program replicates that functionality to the longest chain bullet, and I'm working on the VRF comparison now. Do you think it would be better to put this functionality in the current archive process, or would a separate dedicated process work as well? Also, what does the "best" hash of the state hash mean? This has been a very enlightening discussion, thanks a bunch for taking time out of your day to respond!

**Reply by garethtdavies** [reply]
I think ideally, in the archive process, labelling as such in the database such that it can be queried directly.

The comparing hashes of state hashes is for breaking the tie at the current tip when there is more than 1 block from the same producer (identical ones will simply be rejected). This is extremely common, with most producers running multiple block producing nodes and basically, this is so everyone can agree on the current tip. It was implemented here https://github.com/MinaProtocol/mina/pull/6983, as not 100% sure how this comparison is made.

For the VRF comparison, you might be interested in https://github.com/zkvalidator/mina-vrf-rs as this is in Rust.
